File: tools/HME/proxy/service_registry.js
'use strict';
const path = require('path');
const { loadJsonc } = require('./config_loader');

const PROJECT_ROOT = process.env.PROJECT_ROOT || path.resolve(__dirname, '..', '..', '..');
const FALLBACK_ROOT = path.resolve(__dirname, '..', '..', '..');
const REGISTRY_PATH = (() => {
  const candidate = path.join(PROJECT_ROOT, 'tools', 'HME', 'config', 'services.json');
  try {
    require('fs').accessSync(candidate);
    return candidate;
  } catch (_) {
    return path.join(FALLBACK_ROOT, 'tools', 'HME', 'config', 'services.json');
  }
})();

let _services = null;

function services() {
  if (_services) return _services;
  const data = loadJsonc(REGISTRY_PATH);
  if (!Array.isArray(data.services)) throw new Error(`${REGISTRY_PATH}: services must be a list`);
  _services = data.services;
  return _services;
}

function service(id) {
  const hit = services().find((s) => s && s.id === id);
  if (!hit) throw new Error(`unknown HME service: ${id}`);
  return hit;
}

function servicePort(id, env = process.env) {
  const spec = service(id);
  const raw = env[spec.env_port];
  const value = raw === undefined || raw === null || raw === '' ? spec.default_port : raw;
  const n = Number(value);
  if (!Number.isInteger(n) || n < 1 || n > 65535) {
    throw new Error(`${spec.env_port || id}="${value}" is not a valid port (1-65535)`);
  }
  return n;
}

function serviceHost(id) {
  return service(id).host || '127.0.0.1';
}

function servicePath(id) {
  const p = service(id).health_path || '/health';
  return p.startsWith('/') ? p : `/${p}`;
}

function serviceUrl(id, opts = {}) {
  const p = opts.path || servicePath(id);
  return `http://${serviceHost(id)}:${servicePort(id)}${p.startsWith('/') ? p : `/${p}`}`;
}

function serviceEnabled(id, env = process.env) {
  const rule = service(id).enabled_when;
  if (!rule) return true;
  if (rule.unless_env && env[rule.unless_env] === String(rule.unless_value || '1')) return false;
  return Boolean(rule.env && Array.isArray(rule.in) && rule.in.map(String).includes(String(env[rule.env] || '')));
}

function servicePidLabel(id) {
  const spec = service(id);
  return spec.pid_label || spec.id;
}

function supervisedChildren(parentId, opts = {}) {
  return services().filter((spec) => {
    if (!spec || spec.supervised_by !== parentId) return false;
    if (opts.requiredOnly && !spec.required) return false;
    return serviceEnabled(spec.id, opts.env || process.env);
  });
}

function bundleServices(parentId, opts = {}) {
  return [service(parentId), ...supervisedChildren(parentId, opts)];
}

function bundleProcessPatterns(parentId) {
  const out = [];
  for (const spec of bundleServices(parentId)) {
    for (const pat of spec.process_patterns || []) {
      if (pat && !out.includes(pat)) out.push(pat);
    }
  }
  return out;
}

function bundlePidLabels(parentId) {
  return bundleServices(parentId).map((spec) => spec.pid_label || spec.id);
}

module.exports = {
  REGISTRY_PATH,
  services,
  service,
  servicePort,
  serviceHost,
  servicePath,
  serviceUrl,
  serviceEnabled,
  servicePidLabel,
  supervisedChildren,
  bundleServices,
  bundleProcessPatterns,
  bundlePidLabels,
};

File: tools/HME/scripts/service_registry.py
"""Single source of truth for HME service metadata."""
from __future__ import annotations

import os
from pathlib import Path
from typing import Any

from jsonc import load_jsonc


PROJECT_ROOT = Path(
    os.environ.get("PROJECT_ROOT")
    or os.environ.get("CLAUDE_PROJECT_DIR")
    or Path(__file__).resolve().parents[3]
)
REGISTRY_PATH = PROJECT_ROOT / "tools" / "HME" / "config" / "services.json"


def load_services(root: Path | None = None) -> list[dict[str, Any]]:
    path = (root or PROJECT_ROOT) / "tools" / "HME" / "config" / "services.json"
    data = load_jsonc(path)
    services = data.get("services", [])
    if not isinstance(services, list):
        raise ValueError(f"{path}: services must be a list")
    return services


def service_map(root: Path | None = None) -> dict[str, dict[str, Any]]:
    return {str(s.get("id")): s for s in load_services(root)}


def service_enabled(service: dict[str, Any], env: dict[str, str] | None = None) -> bool:
    rule = service.get("enabled_when")
    if not rule:
        return True
    env = env or os.environ
    unless_env = rule.get("unless_env")
    if unless_env and env.get(str(unless_env)) == str(rule.get("unless_value", "1")):
        return False
    name = rule.get("env")
    allowed = {str(v) for v in rule.get("in", [])}
    return bool(name and env.get(str(name)) in allowed)


def service_url(service: dict[str, Any], env: dict[str, str] | None = None) -> str:
    env = env or os.environ
    host = service.get("host", "127.0.0.1")
    port = env.get(str(service.get("env_port") or ""), service.get("default_port"))
    path = str(service.get("health_path") or "/health")
    if not path.startswith("/"):
        path = "/" + path
    return f"http://{host}:{port}{path}"


def heartbeat_path(service: dict[str, Any], root: Path | None = None) -> Path:
    root = root or PROJECT_ROOT
    rel = service.get("heartbeat_file")
    if not rel:
        raise ValueError(f"{service.get('id')}: missing heartbeat_file")
    return root / str(rel)


def universal_pulse_http_probes(root: Path | None = None) -> list[dict[str, Any]]:
    probes: list[dict[str, Any]] = []
    for service in load_services(root):
        if service.get("kind") != "http":
            continue
        if not service_enabled(service):
            continue
        probes.append({
            "name": service["id"],
            "url": service_url(service),
            "timeout_sec": service.get("timeout_sec", 3),
            "required": bool(service.get("required", True)),
        })
    return probes
